components/org.wso2.carbon.crypto.provider.hsm/src/main/java/org/wso2/carbon/crypto/provider/hsm/HSMBasedExternalCryptoProvider.java

```diff
@@ -0,0 +1,675 @@
+            int tagPos = encryptedData.length - (int) (gcmParams.ulTagBits) / 8;
+            byte[] cipherData = subArray(encryptedData, 0, tagPos);
+            byte[] authTag = subArray(encryptedData, tagPos, (int) (gcmParams.ulTagBits) / 8);
+            return new HybridEncryptionOutput(cipherData, encryptedKey, gcmParams.pAAD,
+                    authTag, gcmParameterSpec);
+        } else if (paramObject instanceof InitializationVectorParameters) {
+            IvParameterSpec ivParameterSpec = new IvParameterSpec(((InitializationVectorParameters)
+                    paramObject).getInitializationVector());
+            return new HybridEncryptionOutput(encryptedData, encryptedKey, ivParameterSpec);
+        } else {
+            String errorMessage = String.format("Invalid / Unsupported parameter specification for '%s' symmetric " +
+                    "encryption algorithm.", symmetricMechanism.getName());
+            throw new CryptoException(errorMessage);
+        }
+    }
+
+    protected byte[] encryptSymmetricKey(SecretKey encryptionKey, String asymmetricAlgorithm, CryptoContext cryptoContext,
+                                         CertificateInfo certificateInfo) throws CryptoException {
+
+        if (log.isDebugEnabled()) {
+            log.debug(String.format("Encrypting generated '%s' symmetric key for hybrid encryption with %s asymmetric " +
+                            "algorithm and public certificate : %s", encryptionKey.getClass().getName(),
+                    asymmetricAlgorithm, certificateInfo));
+        }
+        if (encryptionKey instanceof AESSecretKey) {
+            return encrypt(((AESSecretKey) encryptionKey).getValue().getByteArrayValue(),
+                    asymmetricAlgorithm, null, cryptoContext, certificateInfo);
+        } else if (encryptionKey instanceof DESSecretKey) {
+            return encrypt(((DESSecretKey) encryptionKey).getValue().getByteArrayValue(),
+                    asymmetricAlgorithm, null, cryptoContext, certificateInfo);
+        } else if (encryptionKey instanceof DES2SecretKey) {
+            return encrypt(((DES2SecretKey) encryptionKey).getValue().getByteArrayValue(),
+                    asymmetricAlgorithm, null, cryptoContext, certificateInfo);
+        } else if (encryptionKey instanceof DES3SecretKey) {
+            return encrypt(((DES3SecretKey) encryptionKey).getValue().getByteArrayValue(),
+                    asymmetricAlgorithm, null, cryptoContext, certificateInfo);
+        } else {
+            String errorMessage = String.format("Symmetric encryption key instance '%s' provided for hybrid " +
+                    "encryption is not supported by the provider", encryptionKey.getClass().getName());
+            throw new CryptoException(errorMessage);
+        }
+    }
+
+    protected byte[] subArray(byte[] byteArray, int beginIndex, int length) {
```

Comment:
Why this method is protected. Seems it is only used in this class?

Reply:
To support utility method inheritance in case of extending this class.